### PR TITLE
[macos] fix binding interface and build

### LINKED ISSUE #6339: OpenThread Daemon error macOS
When building posix Deamon on macOS an implicit conversion error is thrown.

running `make -f src/posix/Makefile-posix DAEMON=1 ` will throw following error:
`/../src/posix/platform/daemon.cpp:75:12: error: 
      implicit conversion loses integer precision: 'ssize_t' (aka 'long') to 'int'
      [-Werror,-Wshorten-64-to-32]
    rval = write(sSessionSocket, buf, (size_t)static_cast<size_t>(rval));`

1. Git commit id: 1236839
2. IEEE 802.15.4 hardware platform: nRF52840
3. Build steps: `./bootstrap` and `make -f src/posix/Makefile-posix DAEMON=1`
4. Network topology: -

**Expected behavior**
The Makefile should run and build the daemon without throwing errors

**Console/log output**
`/../src/posix/platform/daemon.cpp:75:12: error: 
      implicit conversion loses integer precision: 'ssize_t' (aka 'long') to 'int'
      [-Werror,-Wshorten-64-to-32]
    rval = write(sSessionSocket, buf, (size_t)static_cast<size_t>(rval));`

**Additional context**
You could fix this error by explicitly converting the value for the rval variable e.g.:
`real = (int)write(sSessionSocket, buf, (size_t)static_cast<size_t>(rval));`
in daemon.cpp line 75

### PR DESCRIPTION
**Pull request by bukepo**
This commit fixes binding interface on macOS and also build errors on
macOS. The pty check is also added on macOS.

Fixes #6339

### CODE REVIEW

File: .github/workflows/posix.yml [hunk 1]
```diff
@@ -277,53 +278,35 @@ jobs:
...
 
   upload-coverage:
     needs:
     - expects-linux
     - posix-cli
     - posix-ncp
     - posix-ncp-rcp-migrate
-    - posix-pty
-    - posix-pty-daemon
+    - pty-linux
```
Comment (simonlingoogle) [inline .github/workflows/posix.yml:309]:
```suggestion
    - pty-linux
    - pty-macos
```

File: script/check-posix-pty [hunk 2]
```diff
@@ -118,11 +121,15 @@ expect "US"
...
 send "panid 0xface\r\n"
 expect "Done"
+send "routerselectionjitter 1\r\n"
+expect "Done"
 send "ifconfig up\r\n"
 expect "Done"
 send "thread start\r\n"
 expect "Done"
-sleep 5
+send "state leader\r\n"
```
Comment (simonlingoogle) [inline script/check-posix-pty:130]:
```suggestion
send "state leader\r\n"
```
why is this necessary ? Node should become leader automatically.

File: script/check-posix-pty [hunk 3]
```diff
@@ -155,9 +163,15 @@ EOF
...
 
-    extaddr=$(grep -azoP 'extaddr[\r\n]+\K[0-9a-z]{16}' $OT_OUTPUT | tr -d '\0')
+    extaddr=$(grep -ao -A +1 'extaddr' $OT_OUTPUT | tail -n1 | tr -d '\r\n\0')
     echo "Extended address is: ${extaddr}"
 
+    prefix=$(grep -ao 'Mesh Local Prefix: [0-9a-f:]\+' $OT_OUTPUT | cut -d: -f2- | tr -d ' \r\n')
+    LEADER_ALOC="${prefix}ff:fe00:fc00"
+    # Retrievie test resource through application CoAP
+
+    [[ $OSTYPE == "linux-gnu"* ]] || return 0
```
Comment (simonlingoogle) [inline script/check-posix-pty:173]: why we don't do following tests in macOS ?

File: .github/workflows/posix.yml (hunk 1 above)
Comment (bukepo) [reply .github/workflows/posix.yml:309]: I failed to get code coverage on macOS yet.

File: script/check-posix-pty (hunk 3 above)
Comment (bukepo) [reply script/check-posix-pty:173]: see https://github.com/openthread/openthread/issues/6363

Comment (simonlingoogle) [reply script/check-posix-pty:197]: Thanks

File: .github/workflows/posix.yml (hunk 1 above)
Comment (simonlingoogle) [reply .github/workflows/posix.yml:309]: Then it would be no difference whether or not `pty-macos` is added here. I am fine to both way.

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

Thanks

**Comment by bukepo** [comment]
There are two checks renamed, but it seems the CI is still waiting for their responses. @jwhui

File: script/check-posix-pty (hunk 3 above)
Comment (jwhui) [reply script/check-posix-pty:197]: Add comment with link to issue here?

File: script/check-posix-pty [hunk 4]
```diff
@@ -140,25 +176,27 @@ send "coap resource TestResource\r\n"
...
-        sleep 5
-        if grep -q leader $OT_OUTPUT; then
-            break
-        else
-            echo 'Still waiting for leader'
-        fi
-    done
+    timeout_run 5 wait_for_leader
 
     netstat -an | grep -q 5683 || die 'Application CoAP port is not available!'
```
Comment (jwhui) [inline script/check-posix-pty:188]: We start coap after node becomes leader. Is it possible for this check to occur before `coap start` is complete?

Comment (bukepo) [reply script/check-posix-pty:191]: added a sleep before.

File: script/check-posix-pty (hunk 3 above)
Comment (bukepo) [reply script/check-posix-pty:200]: added

File: .github/workflows/posix.yml [hunk 5]
```diff
@@ -245,18 +245,21 @@ jobs:
         name: cov-posix-ncp-rcp-migrate
         path: tmp/coverage.info
 
-  posix-pty:
+  pty-linux:
```
Comment (jwhui) [inline .github/workflows/posix.yml:248]: Can we give the different job instances a more describe name that indicates whether daemon mode is enabled? Currently, the job instances are listed as:

- pty-linux (0)
- pty-linux (1)
- pty-macos (0)
- pty-macos (1)

**Review by jwhui — APPROVED** [review]
